### PR TITLE
Add stable code symbol read model

### PR DESCRIPTION
**Pull request by kumanday**
## Summary

Adds stable logical symbol identity and read-model primitives for the Code Graph while preserving revision-bound `symbol_id` provenance and real checked-out `commit_sha` values.

## Changes

- Add AST container-chain extraction on symbol records, including Rust `impl` owner chains, lexical parent chains, and trait-impl qualifiers, and pass container chains into memory persistence.
- Persist `symbol_key`, duplicate ordinal suffixes, container symbol IDs, source/target symbol-key edge endpoints, and normalized edge confidence.
- Add memory read-model helpers for symbol detail, deduplicated span containment, bounded self-contained neighborhoods, and revision-pair comparison by `symbol_key`.
- Keep qualified syntactic call hints unresolved instead of resolving by basename, drop source-less neighborhood edges explicitly, and isolate dirty worktree code-intel rows with `worktree_dirty` while keeping `commit_sha` as the real HEAD.
- Keep code graph read helpers read-only: symbol detail, span containment, neighborhoods, and revision comparisons no longer create or migrate the DuckDB index while serving lookups.
- Schema-gate code graph reads on legacy indexes that predate `symbol_key`/container/edge-key columns, returning empty read results instead of mutating or failing on missing columns.
- Split qualified Rust impl owners such as `m::Widget` into container-chain components so impl methods align with the indexed owner hierarchy.
- Persist the full parsed symbol set for persistent code-intel ingest so edge endpoint resolution is not capped by the artifact display limit.
- Add focused tests for stable identity, duplicates, containers, nested chains, impl parent linking, trait-impl owner links, edge resolution, traversal bounds, read-only lookups, legacy-index schema gating, dirty revision isolation, and revision diff classification/truncation.

### Test output

```text
cargo test-system-duckdb rust_qualified_impl_owners_split_into_container_chain_parts -- --nocapture
result: 1 passed; 0 failed

cargo test-system-duckdb opensymphony_code_intel::tests -- --nocapture
result: 34 passed; 0 failed

cargo fmt --check
result: passed

git diff --check
result: passed

cargo clippy --workspace --all-targets -- -D warnings
result: passed

cargo test-system-duckdb code_intel_read_helpers_schema_gate_legacy_index -- --nocapture
result: 1 passed; 0 failed

cargo test-system-duckdb memory_ingest_code_intel_limit_does_not_cap_persisted_symbols -- --nocapture
result: 1 passed; 0 failed

cargo test-system-duckdb code_intel_read_helpers_do_not_create_index -- --nocapture
result: 1 passed; 0 failed

cargo test-system-duckdb code_intel_impl_methods_link_to_owner_symbol_by_chain -- --nocapture
result: 1 passed; 0 failed

cargo test-system-duckdb code_intel_ -- --nocapture
result: 22 passed; 0 failed

cargo test-system-duckdb opensymphony_code_intel::tests -- --nocapture
result: 33 passed; 0 failed

cargo test-system-duckdb --test memory
result: 46 passed; 0 failed

cargo fmt --check
result: passed

git diff --check
result: passed

cargo clippy --workspace --all-targets -- -D warnings
result: passed
```

### Verification

Reproduced the initial gap before editing: `SymbolRecord` had no container fields, `code_symbols` had no `symbol_key`, and persistence inserted `container_symbol_id` as `NULL`. The tests now verify `symbol_id` changes across content/span changes while `symbol_key` remains stable, duplicate logical symbols receive `#2` suffixes, nested and Rust impl container chains are returned, qualified Rust impl owners split into container-chain parts before method key generation, trait impl methods include the trait qualifier in stable keys, nested Rust impl methods preserve lexical parents, trait impl methods link to owner symbols even inside lexical containers, persistent ingest keeps all parsed symbols available for edge resolution despite a low artifact limit, code graph read helpers do not create or migrate the index, legacy pre-code-graph indexes return empty read results without adding columns, syntactic/qualified edge targets stay honest, neighborhoods omit source-less edges and report truncation, dirty worktree rows keep real `commit_sha` values without polluting clean revision comparisons, and revision comparison returns added/removed/modified symbols without false truncation on exact-sized pages.

## Checklist

- [x] Tests pass (`cargo test`)
- [x] Code is formatted (`cargo fmt`)
- [x] Clippy is clean (`cargo clippy`)
- [x] Documentation updated (if behavior changed)
- [x] `AGENTS.md` updated (if repo knowledge changed)
- [x] Evidence section filled out (for substantive changes)

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Documentation
- [ ] Other:

## Breaking changes

None. The existing revision-bound `symbol_id` remains revision-scoped; `symbol_key` is added beside it as stable logical identity, and persisted `commit_sha` values remain the checked-out revision.

## Related issues

COE-532

## Additional notes

The requested `cargo test-system-duckdb --test code_intel` target does not exist in this repo; code-intel coverage is under `opensymphony_code_intel::tests`, which passed.

### CODE REVIEW

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-memory/src/index.rs [hunk 1]
```diff
@@ -729,6 +732,595 @@ pub fn persist_code_intel_documents(
...
+    name: &str,
+) -> Option<&'a PreparedCodeSymbol<'a>> {
+    let mut matches = symbols.iter().filter(|symbol| symbol.symbol.name == name);
+    let first = matches.next()?;
+    matches.next().is_none().then_some(first)
+}
+
+fn normalize_edge_confidence(input: &str, target_resolved: bool) -> &'static str {
+    if target_resolved {
+        return "exact";
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in d2fc9e4: syntactic AST captures now stay syntactic unless the input confidence itself is exact and the target resolves.

File: crates/opensymphony-memory/src/index.rs [hunk 2]
```diff
@@ -568,33 +577,24 @@ pub fn persist_code_intel_documents(
...
+                        prepared.symbol_key,
                         batch.repo_id,
                         batch.commit_sha,
                         batch.worktree_dirty,
                         path,
                         document.language,
                         symbol.kind,
                         symbol.name,
-                        Option::<String>::None,
+                        prepared.container_symbol_id,
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs:678]: Fixed in d2fc9e4: persisted symbols now store the raw container chain, so partial symbol indexes retain container context even without parent rows.

File: crates/opensymphony-memory/src/index.rs [hunk 3]
```diff
@@ -729,6 +732,595 @@ pub fn persist_code_intel_documents(
...
+    let mut base_key_counts = BTreeMap::<String, usize>::new();
+    let mut prepared = document
+        .symbols
+        .iter()
+        .map(|symbol| {
+            let symbol_id = code_row_id(&[
+                repo_id,
+                path,
+                &document.content_sha256,
+                &document.parser_version,
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs:850]: Fixed in d2fc9e4: symbol_id now includes revision provenance, preserving unchanged symbol rows for base/head comparisons while keeping symbol_key stable.

File: crates/opensymphony-memory/src/index.rs [hunk 4]
```diff
@@ -729,6 +732,595 @@ pub fn persist_code_intel_documents(
...
+        start_line: row.get::<_, i64>(7)? as usize,
+        start_col: row.get::<_, i64>(8)? as usize,
+        end_line: row.get::<_, i64>(9)? as usize,
+        end_col: row.get::<_, i64>(10)? as usize,
+    })
+}
+
+fn symbol_contains_point(symbol: &CodeSymbolRecord, line: usize, column: usize) -> bool {
+    (symbol.start_line < line || (symbol.start_line == line && symbol.start_col <= column))
+        && (symbol.end_line > line || (symbol.end_line == line && symbol.end_col >= column))
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in d2fc9e4: span containment now treats end columns as exclusive and has regression coverage for endpoint lookups.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-memory/src/index.rs [hunk 5]
```diff
@@ -729,6 +742,614 @@ pub fn persist_code_intel_documents(
...
+    limit: usize,
+) -> Result<Vec<CodeSymbolRecord>, MemoryError> {
+    let connection = open_index(config)?;
+    migrate_index(&connection).map_err(|source| MemoryError::DuckDb {
+        path: config.index_path.clone(),
+        source,
+    })?;
+    let mut statement = connection
+        .prepare(
+            "SELECT symbol_id, symbol_key, repo_id, commit_sha, path, language, kind, name, container_symbol_id, container_chain, start_line, start_col, end_line, end_col, start_byte, end_byte, snippet_sha256, freshness FROM code_symbols WHERE repo_id = ? AND path = ? AND freshness = 'current' ORDER BY start_line, start_col, end_line DESC, end_col DESC, symbol_key",
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in 032ab37: span containment now collapses current rows by symbol_key so unchanged base/head rows return one logical symbol.

File: crates/opensymphony-code-intel/src/lib.rs [hunk 6]
```diff
@@ -1512,9 +1518,47 @@ fn run_query_pack(
...
 
+fn assign_symbol_containers(symbols: &mut [SymbolRecord]) {
+    for index in 0..symbols.len() {
+        let parent_index = symbols
+            .iter()
+            .enumerate()
+            .filter(|(candidate_index, candidate)| {
+                *candidate_index != index
+                    && span_strictly_contains(&candidate.span, &symbols[index].span)
+            })
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-code-intel/src/lib.rs:1546]: Fixed in 032ab37: Rust methods inside impl blocks now use the impl owner as their container chain when no captured parent symbol exists.

File: crates/opensymphony-memory/src/index.rs [hunk 7]
```diff
@@ -729,6 +742,614 @@ pub fn persist_code_intel_documents(
...
+
+fn symbol_contains_span(symbol: &CodeIntelSymbolInput, start_byte: usize, end_byte: usize) -> bool {
+    symbol.start_byte <= start_byte && symbol.end_byte >= end_byte
+}
+
+fn edge_target_name(target_hint: &str) -> Option<&str> {
+    let before_call = target_hint.split_once('(').map_or(target_hint, |(name, _)| name);
+    before_call
+        .split([':', '.'])
+        .rfind(|part| !part.is_empty())
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in 032ab37: qualified call hints now remain unresolved instead of resolving by basename to an unrelated local symbol.

File: crates/opensymphony-memory/src/index.rs [hunk 8]
```diff
@@ -729,6 +742,614 @@ pub fn persist_code_intel_documents(
...
+}
+
+fn query_symbols_for_revision(
+    connection: &Connection,
+    repo_id: &str,
+    revision: &str,
+) -> Result<BTreeMap<String, CodeSymbolRecord>, MemoryError> {
+    let mut statement = connection
+        .prepare(
+            "SELECT symbol_id, symbol_key, repo_id, commit_sha, path, language, kind, name, container_symbol_id, container_chain, start_line, start_col, end_line, end_col, start_byte, end_byte, snippet_sha256, freshness FROM code_symbols WHERE repo_id = ? AND commit_sha = ? ORDER BY symbol_key, indexed_at DESC, symbol_id",
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in 032ab37: dirty worktree code-intel rows are stored under a separate dirty revision key, so clean revision comparisons ignore uncommitted snippets.

File: crates/opensymphony-memory/src/index.rs [hunk 9]
```diff
@@ -729,6 +742,614 @@ pub fn persist_code_intel_documents(
...
+                            truncated = true;
+                            continue;
+                        }
+                        if let Some(symbol) = query_code_symbol_by_key(&connection, adjacent, true)? {
+                            next_frontier.insert(adjacent.to_string());
+                            symbols.insert(adjacent.to_string(), symbol);
+                        }
+                    }
+                }
+                edges.insert(edge.edge_id.clone(), edge);
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs:1188]: Fixed in 032ab37: bounded neighborhoods now skip edges whose resolved endpoints were omitted by the record cap.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-code-intel/src/lib.rs [hunk 10]
```diff
@@ -1512,9 +1520,81 @@ fn run_query_pack(
...
+    let name = owner
+        .utf8_text(source)?
+        .split('<')
+        .next()
+        .unwrap_or("")
+        .trim();
+    if name.is_empty() {
+        Ok((None, Vec::new()))
+    } else {
+        Ok((Some(name.to_string()), vec![name.to_string()]))
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-code-intel/src/lib.rs]: Fixed in f09d377: Rust trait impl methods now include the trait qualifier in their container chain, so Widget/Display/fmt and Widget/Debug/fmt keep distinct stable keys.

File: crates/opensymphony-memory/src/index.rs [hunk 11]
```diff
@@ -729,6 +744,629 @@ pub fn persist_code_intel_documents(
...
+    let parent_name = child.container_chain.last()?;
+    let parent_chain = &child.container_chain[..child.container_chain.len().saturating_sub(1)];
+    symbols
+        .iter()
+        .enumerate()
+        .filter(|(candidate_index, candidate)| {
+            *candidate_index != child_index
+                && candidate.symbol.name == *parent_name
+                && candidate.symbol.container_chain == parent_chain
+                && symbol_contains_span(candidate.symbol, child.start_byte, child.end_byte)
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in f09d377: container parent lookup keeps span-based resolution first and falls back to a unique chain/name owner match for Rust impl methods outside the type span.

File: crates/opensymphony-code-intel/src/lib.rs [hunk 12]
```diff
@@ -1512,9 +1520,81 @@ fn run_query_pack(
...
+                        .saturating_sub(candidate.span.start_byte),
+                    std::cmp::Reverse(candidate.span.start_byte),
+                )
+            })
+            .map(|(candidate_index, _)| candidate_index);
+
+        if let Some(parent_index) = parent_index {
+            let parent = symbols[parent_index].clone();
+            let mut chain = parent.container_chain;
+            chain.push(parent.name.clone());
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-code-intel/src/lib.rs:1561]: Fixed in f09d377: container assignment now processes wider parent symbols first, so nested symbols receive complete root-to-leaf chains regardless of query emission order.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-memory/src/index.rs [hunk 13]
```diff
@@ -765,6 +1408,16 @@ fn code_row_id(parts: &[&str]) -> String {
     sha256_hex(&parts.join("\u{1f}"))
 }
 
+fn code_revision_key(commit_sha: Option<&str>, worktree_dirty: bool) -> Option<String> {
+    commit_sha.map(|sha| {
+        if worktree_dirty {
+            format!("{sha}+dirty")
+        } else {
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in 432cdf5: dirty ingests now keep commit_sha as the checked-out HEAD, keep dirty row identity separate, and clean revision comparison filters dirty rows without rewriting the commit value.

File: crates/opensymphony-memory/src/index.rs [hunk 14]
```diff
@@ -729,6 +744,634 @@ pub fn persist_code_intel_documents(
...
+        }
+    }
+
+    prepared
+}
+
+fn container_symbol_index(symbols: &[PreparedCodeSymbol<'_>], child_index: usize) -> Option<usize> {
+    let child = symbols[child_index].symbol;
+    let parent_name = child.container_chain.last()?;
+    let parent_chain = &child.container_chain[..child.container_chain.len().saturating_sub(1)];
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs:895]: Fixed in 432cdf5: trait impl method parent resolution now falls back to the owner symbol for Widget/Display-style chains, with regression coverage for the owner link.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-memory/src/index.rs [hunk 15]
```diff
@@ -729,6 +764,666 @@ pub fn persist_code_intel_documents(
...
+        max_records,
+        truncated,
+    }))
+}
+
+fn edge_endpoints_present(edge: &CodeEdgeRecord, symbols: &BTreeMap<String, CodeSymbolRecord>) -> bool {
+    [edge.source_symbol_key.as_deref(), edge.target_symbol_key.as_deref()]
+        .into_iter()
+        .flatten()
+        .all(|key| symbols.contains_key(key))
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in ca740a5: neighborhoods now require a present source symbol before returning an edge, while still preserving explicitly unresolved target edges.

File: crates/opensymphony-code-intel/src/lib.rs [hunk 16]
```diff
@@ -1512,9 +1520,103 @@ fn run_query_pack(
...
+        std::cmp::Reverse(
+            symbols[*index]
+                .span
+                .end_byte
+                .saturating_sub(symbols[*index].span.start_byte),
+        )
+    });
+    for index in indexes {
+        if !symbols[index].container_chain.is_empty() {
+            continue;
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-code-intel/src/lib.rs]: Fixed in ca740a5: Rust impl methods now prepend lexical parent chains to precomputed impl owner chains, preserving scoped Widget/run identities.

File: crates/opensymphony-memory/src/index.rs [hunk 17]
```diff
@@ -729,6 +764,666 @@ pub fn persist_code_intel_documents(
...
+    let head = query_symbols_for_revision(&connection, repo_id, head_revision)?;
+    let mut keys = base.keys().cloned().collect::<BTreeSet<_>>();
+    keys.extend(head.keys().cloned());
+    let mut diffs = Vec::new();
+    let mut truncated = false;
+
+    for key in keys {
+        if diffs.len() >= max_records {
+            truncated = true;
+            break;
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in ca740a5: revision comparison now checks the record cap only after a key produces an actual added/removed/modified diff.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-memory/src/index.rs [hunk 18]
```diff
@@ -729,6 +764,675 @@ pub fn persist_code_intel_documents(
...
+
+pub fn code_symbol_detail(
+    config: &MemoryConfig,
+    symbol_key: &str,
+) -> Result<Option<CodeSymbolRecord>, MemoryError> {
+    let connection = open_index(config)?;
+    migrate_index(&connection).map_err(|source| MemoryError::DuckDb {
+        path: config.index_path.clone(),
+        source,
+    })?;
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in 945111d: code graph read helpers now open the existing DuckDB index read-only and return empty results when the index is absent instead of migrating during lookup.

File: crates/opensymphony-memory/src/index.rs [hunk 19]
```diff
@@ -620,6 +652,7 @@ pub fn persist_code_intel_documents(
         }
 
         for edge in &document.edges {
+            let resolved = resolve_code_edge(edge, &prepared_symbols);
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs:705]: Fixed in 945111d: persistent code-intel ingest now stores the full parsed symbol set, so edge endpoint resolution is not capped by the artifact limit.

File: crates/opensymphony-memory/src/index.rs [hunk 20]
```diff
@@ -729,6 +764,675 @@ pub fn persist_code_intel_documents(
...
+
+fn trait_impl_owner_symbol_index(
+    symbols: &[PreparedCodeSymbol<'_>],
+    child_index: usize,
+) -> Option<usize> {
+    let child = symbols[child_index].symbol;
+    if child.container_chain.len() < 2 {
+        return None;
+    }
+    let owner_name = child.container_chain.first()?;
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs]: Fixed in 945111d: trait impl method parent fallback now selects the implementing type immediately before the trait segment in lexical container chains.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-memory/src/index.rs [hunk 21]
```diff
@@ -248,6 +248,13 @@ fn open_index_read_only(config: &MemoryConfig) -> Result<Connection, MemoryError
     })
 }
 
+fn open_existing_index_read_only(config: &MemoryConfig) -> Result<Option<Connection>, MemoryError> {
+    if !config.index_path.exists() {
+        return Ok(None);
+    }
+    open_index_read_only(config).map(Some)
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-memory/src/index.rs:255]: Fixed in a6c3455: code graph read helpers now schema-gate legacy indexes in read-only mode and return empty results instead of querying missing columns or migrating during lookup.

**Comment by kumanday** [comment]
@codex review

File: crates/opensymphony-code-intel/src/lib.rs [hunk 22]
```diff
@@ -1512,9 +1520,106 @@ fn run_query_pack(
...
+            Some(trait_name.to_string()),
+            vec![name.to_string(), trait_name.to_string()],
+        ))
+    } else {
+        Ok((Some(name.to_string()), vec![name.to_string()]))
+    }
+}
+
+fn rust_container_part(text: &str) -> Option<&str> {
+    let trimmed = text.split('<').next().unwrap_or(text).trim();
```
[the remark this answers is not in the document]
Comment (kumanday) [reply crates/opensymphony-code-intel/src/lib.rs:1612]: Fixed in cc0abe5: Rust impl owner paths now split qualified owners like m::Widget into container-chain parts, with regression coverage for the extracted method chain.

**Comment by kumanday** [comment]
@codex review
